File: android/src/main/java/expo/modules/spline/ReactNativeSplineView.kt
package expo.modules.spline

import android.content.Context
import design.spline.runtime.SplineView
import design.spline.runtime.SplineEventName
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView

class ReactNativeSplineView(context: Context, appContext: AppContext) : ExpoView(context, appContext) {
  private val onLoad by EventDispatcher()
  private val onSplineEvent by EventDispatcher()

  /// The native Spline view from the SplineRuntime SDK.
  internal val splineView = SplineView(context).also { view ->
    view.layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
    addView(view)
  }

  /// Load a Spline scene from a remote URL and register all event listeners once loaded.
  fun loadUrl(url: String) {
    splineView.loadUrl(url) {
      onLoad(mapOf("url" to url))
      registerEventListeners()
    }
  }

  private fun registerEventListeners() {
    val eventNames = listOf(
      SplineEventName.mouseUp to "mouseUp",
      SplineEventName.mouseDown to "mouseDown",
      SplineEventName.mousePress to "mousePress",
      SplineEventName.mouseHover to "mouseHover",
      SplineEventName.keyUp to "keyUp",
      SplineEventName.keyDown to "keyDown",
      SplineEventName.keyPress to "keyPress",
      SplineEventName.start to "start",
      SplineEventName.lookAt to "lookAt",
      SplineEventName.follow to "follow",
    )

    for ((eventType, eventName) in eventNames) {
      splineView.addEventListener(eventType) { event ->
        onSplineEvent(
          mapOf(
            "event" to eventName,
            "objectName" to event.target.name,
            "objectId" to event.target.uuid,
          )
        )
      }
    }
  }
}

File: ios/ReactNativeSplineView.swift
import ExpoModulesCore
import SplineRuntime
import SwiftUI

// MARK: - ExpoView bridge

/// UIKit view that hosts a SwiftUI SplineView from the SplineRuntime SDK.
class ReactNativeSplineView: ExpoView {
    let onLoad = EventDispatcher()
    let onSplineEvent = EventDispatcher()

    /// Exposed so the module can call SplineController API methods imperatively.
    private(set) var controller = SplineController()

    private var hostingContainer: UIView?

    required init(appContext: AppContext? = nil) {
        super.init(appContext: appContext)
        backgroundColor = .clear
        clipsToBounds = true
    }

    /// Called when the `url` prop changes. Loads the Spline scene from the given URL.
    func loadScene(url: URL) {
        hostingContainer?.removeFromSuperview()
        hostingContainer = nil

        // Fresh controller for each new scene so previous listeners don't leak.
        controller = SplineController()

        let sceneView = SplineSceneView(
            url: url,
            controller: controller,
            onSceneLoaded: {
                [weak self] in
                self?.onLoad(["url": url.absoluteString])
            },
            onEvent: {
                [weak self] eventName, objectName, objectId in
                let payload: [String: Any] = [
                    "event": eventName,
                    "objectName": objectName,
                    "objectId": objectId,
                ]
                self?.onSplineEvent(payload)
            }
        )

        // Bridge SwiftUI → UIKit via UIHostingController.
        let hosting = UIHostingController(rootView: sceneView)
        hosting.view.backgroundColor = .clear
        hosting.view.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        hosting.view.frame = bounds
        addSubview(hosting.view)
        hostingContainer = hosting.view
    }

    override func layoutSubviews() {
        super.layoutSubviews()
        hostingContainer?.frame = bounds
    }
}

// MARK: - SwiftUI wrapper

/// Wraps `SplineView` from SplineRuntime and wires up all Spline event listeners
/// once the scene has finished loading via the phase trailing closure.
private struct SplineSceneView: View {
    let url: URL
    let controller: SplineController
    let onSceneLoaded: () -> Void
    let onEvent: (String, String, String) -> Void

    var body: some View {
        SplineView(sceneFileURL: url, controller: controller) { phase in
            phase.content?.task {
                onSceneLoaded()
                for (eventName, eventType) in [
                    ("mouseUp", SplineEventName.mouseUp),
                    ("mouseDown", SplineEventName.mouseDown),
                    ("mousePress", SplineEventName.mousePress),
                    ("mouseHover", SplineEventName.mouseHover),
                    ("keyUp", SplineEventName.keyUp),
                    ("keyDown", SplineEventName.keyDown),
                    ("keyPress", SplineEventName.keyPress),
                    ("start", SplineEventName.start),
                    ("lookAt", SplineEventName.lookAt),
                    ("follow", SplineEventName.follow),
                ] {
                    controller.addEventListener(eventType) { obj in
                        print("[ReactNativeSpline] event '\(eventName)' on object '\(obj.name)'")
                        onEvent(eventName, obj.name, obj.uuid)
                    }
                }
            }
        }.ignoresSafeArea(.all)
    }
}
